# Stage 3 Challenge

## Preparation

In [2]:
!pip install ipython-sql
import pandas as pd
import sqlite3
%load_ext sql

### Now your turn to create new table to SQL
Use this link as new csv_link

Name it as "chigago_criminal"

In [3]:
# Pandas limits the number of columns displayed.
# The code below is used to show all the columns in the table.
pd.option_context('display.max_columns', None)

In [4]:
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv"
data_3 = pd.read_csv(url)
data_3.head(3)

ID CASE_NUMBER        DATE                     BLOCK IUCR  \
0  3512276    HK587712  2004-08-28        047XX S KEDZIE AVE  890   
1  3406613    HK456306  2004-06-26  009XX N CENTRAL PARK AVE  820   
2  8002131    HT233595  2011-04-04        043XX S WABASH AVE  820   

  PRIMARY_TYPE     DESCRIPTION          LOCATION_DESCRIPTION  ARREST  \
0        THEFT   FROM BUILDING            SMALL RETAIL STORE   False   
1        THEFT  $500 AND UNDER                         OTHER   False   
2        THEFT  $500 AND UNDER  NURSING HOME/RETIREMENT HOME   False   

   DOMESTIC  ...  DISTRICT  WARD  COMMUNITY_AREA_NUMBER  FBICODE X_COORDINATE  \
0     False  ...         9  14.0                   58.0        6    1155838.0   
1     False  ...        11  27.0                   23.0        6    1152206.0   
2     False  ...         2   3.0                   38.0        6    1177436.0   

   Y_COORDINATE  YEAR   LATITUDE  LONGITUDE                       LOCATION  
0     1873050.0  2004  41.807440 -87.703956    (41.8074405, -87.703955849)  
1     1906127.0  2004  41.898280 -87.716406  (41.898279962, -87.716405505)  
2     1876313.0  2011  41.815933 -87.624642  (41.815933131, -87.624642127)  

[3 rows x 21 columns]

In [5]:
data_3.set_index('ID', inplace=True)

### Try Transform this new Data by yourself

# Challenge 1

## Exploration

Lets explore the data first to figure out what we need to do

In [6]:
### Code Start Here, Goodluck
### Initial introduction to the data
data_3.head()

CASE_NUMBER        DATE                     BLOCK IUCR PRIMARY_TYPE  \
ID                                                                             
3512276     HK587712  2004-08-28        047XX S KEDZIE AVE  890        THEFT   
3406613     HK456306  2004-06-26  009XX N CENTRAL PARK AVE  820        THEFT   
8002131     HT233595  2011-04-04        043XX S WABASH AVE  820        THEFT   
7903289     HT133522  2010-12-30      083XX S KINGSTON AVE  840        THEFT   
10402076    HZ138551  2016-02-02           033XX W 66TH ST  820        THEFT   

                            DESCRIPTION          LOCATION_DESCRIPTION  ARREST  \
ID                                                                              
3512276                   FROM BUILDING            SMALL RETAIL STORE   False   
3406613                  $500 AND UNDER                         OTHER   False   
8002131                  $500 AND UNDER  NURSING HOME/RETIREMENT HOME   False   
7903289   FINANCIAL ID THEFT: OVER $300                     RESIDENCE   False   
10402076                 $500 AND UNDER                         ALLEY   False   

          DOMESTIC  BEAT  DISTRICT  WARD  COMMUNITY_AREA_NUMBER FBICODE  \
ID                                                                        
3512276      False   911         9  14.0                   58.0       6   
3406613      False  1112        11  27.0                   23.0       6   
8002131      False   221         2   3.0                   38.0       6   
7903289      False   423         4   7.0                   46.0       6   
10402076     False   831         8  15.0                   66.0       6   

          X_COORDINATE  Y_COORDINATE  YEAR   LATITUDE  LONGITUDE  \
ID                                                                 
3512276      1155838.0     1873050.0  2004  41.807440 -87.703956   
3406613      1152206.0     1906127.0  2004  41.898280 -87.716406   
8002131      1177436.0     1876313.0  2011  41.815933 -87.624642   
7903289      1194622.0     1850125.0  2010  41.743665 -87.562463   
10402076     1155240.0     1860661.0  2016  41.773455 -87.706480   

                               LOCATION  
ID                                       
3512276     (41.8074405, -87.703955849)  
3406613   (41.898279962, -87.716405505)  
8002131   (41.815933131, -87.624642127)  
7903289   (41.743665322, -87.562462756)  
10402076  (41.773455295, -87.706480471)

Looking at the data, we can clean the location value by removing the parenthesis

In [7]:
### lets check for the data dimension
data_3.shape

(533, 20)

the data has 533 rows and 20 columns

Lets explore further

In [8]:
data_3.describe().T

count          mean           std           min  \
BEAT                   533.0  1.187038e+03    660.592263  1.110000e+02   
DISTRICT               533.0  1.119512e+01      6.496631  1.000000e+00   
WARD                   490.0  2.260816e+01     13.080119  1.000000e+00   
COMMUNITY_AREA_NUMBER  490.0  3.754286e+01     21.409514  1.000000e+00   
X_COORDINATE           529.0  1.162876e+06  16501.969976  1.100658e+06   
Y_COORDINATE           529.0  1.886223e+06  30099.803592  1.814512e+06   
YEAR                   533.0  2.008343e+03      4.949400  2.001000e+03   
LATITUDE               529.0  4.184343e+01      0.082798  4.164580e+01   
LONGITUDE              529.0 -8.767782e+01      0.060073 -8.790523e+01   

                                25%           50%           75%           max  
BEAT                   7.110000e+02  1.111000e+03  1.634000e+03  2.535000e+03  
DISTRICT               6.000000e+00  1.000000e+01  1.600000e+01  2.500000e+01  
WARD                   1.200000e+01  2.400000e+01  3.200000e+01  5.000000e+01  
COMMUNITY_AREA_NUMBER  2.300000e+01  3.000000e+01  5.800000e+01  7.700000e+01  
X_COORDINATE           1.151715e+06  1.163235e+06  1.174618e+06  1.204126e+06  
Y_COORDINATE           1.860454e+06  1.891447e+06  1.907711e+06  1.951001e+06  
YEAR                   2.004000e+03  2.008000e+03  2.012000e+03  2.018000e+03  
LATITUDE               4.177256e+01  4.185774e+01  4.190230e+01  4.202118e+01  
LONGITUDE             -8.771817e+01 -8.767673e+01 -8.763468e+01 -8.752822e+01

In [9]:
data_3.describe(include = ['O']).T

count unique                            top freq
CASE_NUMBER            533    533                       HK587712    1
DATE                   533    509                     2012-05-02    3
BLOCK                  533    518           071XX S MICHIGAN AVE    2
IUCR                   533    112                            820   48
PRIMARY_TYPE           533     35                          THEFT  106
DESCRIPTION            533    103                         SIMPLE   59
LOCATION_DESCRIPTION   533     48                         STREET  136
FBICODE                533     23                              6  106
LOCATION               529    525  (41.754592961, -87.741528537)    2

In [10]:
data_3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 533 entries, 3512276 to 1326195
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   CASE_NUMBER            533 non-null    object 
 1   DATE                   533 non-null    object 
 2   BLOCK                  533 non-null    object 
 3   IUCR                   533 non-null    object 
 4   PRIMARY_TYPE           533 non-null    object 
 5   DESCRIPTION            533 non-null    object 
 6   LOCATION_DESCRIPTION   533 non-null    object 
 7   ARREST                 533 non-null    bool   
 8   DOMESTIC               533 non-null    bool   
 9   BEAT                   533 non-null    int64  
 10  DISTRICT               533 non-null    int64  
 11  WARD                   490 non-null    float64
 12  COMMUNITY_AREA_NUMBER  490 non-null    float64
 13  FBICODE                533 non-null    object 
 14  X_COORDINATE           529 non-null    float64
 

There are some data types mismatch, we need to change it to the correct data type, such as Date columns to be converted to 'date' data type

In [11]:
data_3.isna().sum()

CASE_NUMBER               0
DATE                      0
BLOCK                     0
IUCR                      0
PRIMARY_TYPE              0
DESCRIPTION               0
LOCATION_DESCRIPTION      0
ARREST                    0
DOMESTIC                  0
BEAT                      0
DISTRICT                  0
WARD                     43
COMMUNITY_AREA_NUMBER    43
FBICODE                   0
X_COORDINATE              4
Y_COORDINATE              4
YEAR                      0
LATITUDE                  4
LONGITUDE                 4
LOCATION                  4
dtype: int64

Some columns have missing values, we need to fill it with suited values

In [12]:
data_3.duplicated().sum()

0

No duplicated value detected

### Data Cleaning

Lets replace the '(' and ')' characters in the LOCATION column

In [13]:
data_3['LOCATION'] = data_3['LOCATION'].str.replace('(','')
data_3['LOCATION'] = data_3['LOCATION'].str.replace(')','')
print('Done!')

Done!


C:\Users\970232\AppData\Local\Temp\ipykernel_14136\3575529682.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  data_3['LOCATION'] = data_3['LOCATION'].str.replace('(','')
C:\Users\970232\AppData\Local\Temp\ipykernel_14136\3575529682.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  data_3['LOCATION'] = data_3['LOCATION'].str.replace(')','')


Lets change the mismatched data type to the correct one

In [14]:
### Code below used to convert the columns
data_3['DATE'] = data_3['DATE'].apply(pd.to_datetime)

In [15]:
data_3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 533 entries, 3512276 to 1326195
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   CASE_NUMBER            533 non-null    object        
 1   DATE                   533 non-null    datetime64[ns]
 2   BLOCK                  533 non-null    object        
 3   IUCR                   533 non-null    object        
 4   PRIMARY_TYPE           533 non-null    object        
 5   DESCRIPTION            533 non-null    object        
 6   LOCATION_DESCRIPTION   533 non-null    object        
 7   ARREST                 533 non-null    bool          
 8   DOMESTIC               533 non-null    bool          
 9   BEAT                   533 non-null    int64         
 10  DISTRICT               533 non-null    int64         
 11  WARD                   490 non-null    float64       
 12  COMMUNITY_AREA_NUMBER  490 non-null    float64       


Done, lets fill NA values

In [16]:
to_fill = ['WARD', 'COMMUNITY_AREA_NUMBER', 'X_COORDINATE', 'Y_COORDINATE', 'LATITUDE', 'LONGITUDE', 'LOCATION']

for i in to_fill:
    data_3[i].fillna('Unknown', inplace = True)

In [17]:
data_3.isna().sum()

CASE_NUMBER              0
DATE                     0
BLOCK                    0
IUCR                     0
PRIMARY_TYPE             0
DESCRIPTION              0
LOCATION_DESCRIPTION     0
ARREST                   0
DOMESTIC                 0
BEAT                     0
DISTRICT                 0
WARD                     0
COMMUNITY_AREA_NUMBER    0
FBICODE                  0
X_COORDINATE             0
Y_COORDINATE             0
YEAR                     0
LATITUDE                 0
LONGITUDE                0
LOCATION                 0
dtype: int64

Done, now the cleaning is finished lets store the dataframe into a new copy fit for assessment

In [18]:
df_by_elt = data_3.copy()
print("Done!")

Done!


## Challenge 2

If in stage 1 (data_1) we import data then we use python.

Try importing processed data in SQL. Try your own SQL. No manipulating in python.

Lets import a fresh file

In [19]:
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv"
data_4 = pd.read_csv(url)
data_4.set_index('ID', inplace=True)

In [20]:
# define your database name
db = 'chicago_crime.db'

# connect to local database or create new if not exist
con = sqlite3.connect(db)
cur = con.cursor()

In [21]:
table_name_crime = 'chicago_crime'
data_4.to_sql(table_name_crime, con=con, index=True, if_exists='replace')

print("Hello :)")

Hello :)


In [22]:
%sql sqlite:///chicago_crime.db

In [23]:
%%sql
SELECT * FROM chicago_crime
LIMIT 5

 * sqlite:///chicago_crime.db
Done.


[(3512276, 'HK587712', '2004-08-28', '047XX S KEDZIE AVE', '890', 'THEFT', 'FROM BUILDING', 'SMALL RETAIL STORE', 0, 0, 911, 9, 14.0, 58.0, '6', 1155838.0, 1873050.0, 2004, 41.8074405, -87.70395585, '(41.8074405, -87.703955849)'),
 (3406613, 'HK456306', '2004-06-26', '009XX N CENTRAL PARK AVE', '820', 'THEFT', '$500 AND UNDER', 'OTHER', 0, 0, 1112, 11, 27.0, 23.0, '6', 1152206.0, 1906127.0, 2004, 41.89827996, -87.71640551, '(41.898279962, -87.716405505)'),
 (8002131, 'HT233595', '2011-04-04', '043XX S WABASH AVE', '820', 'THEFT', '$500 AND UNDER', 'NURSING HOME/RETIREMENT HOME', 0, 0, 221, 2, 3.0, 38.0, '6', 1177436.0, 1876313.0, 2011, 41.81593313, -87.62464213, '(41.815933131, -87.624642127)'),
 (7903289, 'HT133522', '2010-12-30', '083XX S KINGSTON AVE', '840', 'THEFT', 'FINANCIAL ID THEFT: OVER $300', 'RESIDENCE', 0, 0, 423, 4, 7.0, 46.0, '6', 1194622.0, 1850125.0, 2010, 41.74366532, -87.56246276, '(41.743665322, -87.562462756)'),
 (10402076, 'HZ138551', '2016-02-02', '033XX W 66TH ST', '820', 'THEFT', '$500 AND UNDER', 'ALLEY', 0, 0, 831, 8, 15.0, 66.0, '6', 1155240.0, 1860661.0, 2016, 41.7734553, -87.70648047, '(41.773455295, -87.706480471)')]

In [24]:
%%sql
UPDATE chicago_crime
SET LOCATION = REPLACE(LOCATION, "(", "")

 * sqlite:///chicago_crime.db
533 rows affected.


[]

In [25]:
%%sql
UPDATE chicago_crime
SET LOCATION = REPLACE(LOCATION, ")", "")

 * sqlite:///chicago_crime.db
533 rows affected.


[]

The challenge requires us to mimic the same treatment done in python and doing it in sql, so there's data types changes and missing value impute

Lets do data types changes first

In [26]:
#Your Query
your_etl_query = """
SELECT 
    ID, 
    CASE_NUMBER,
    DATE, 
    BLOCK, 
    IUCR, 
    PRIMARY_TYPE,
    DESCRIPTION, 
    LOCATION_DESCRIPTION, 
    CASE WHEN ARREST = 0 THEN "False" ELSE "True" END AS ARREST, 
    CASE WHEN ARREST = 0 THEN "False" ELSE "True" END AS DOMESTIC, 
    BEAT,
    DISTRICT,
    COALESCE(WARD, 'Unknown') AS WARD,
    COALESCE(COMMUNITY_AREA_NUMBER, 'Unknown') AS COMMUNITY_AREA_NUMBER,
    FBICODE, 
    COALESCE(X_COORDINATE, 'Unknown') AS X_COORDINATE,
    COALESCE(Y_COORDINATE, 'Unknown') AS Y_COORDINATE,
    YEAR, 
    COALESCE(LATITUDE, 'Unknown') AS LATITUDE,
    COALESCE(LONGITUDE, 'Unknown') AS LONGITUDE,
    COALESCE(LOCATION, 'Unknown') AS LOCATION
FROM chicago_crime
"""

In [40]:
pd.option_context('display.max_columns', None)
df_your_etl = pd.read_sql(your_etl_query, con=con)
df_your_etl.head()
df_your_etl.shape

(533, 21)

In [42]:
df_your_etl.set_index('ID', inplace = True)

In [43]:
df_your_etl.head()

CASE_NUMBER        DATE                     BLOCK IUCR PRIMARY_TYPE  \
ID                                                                             
3512276     HK587712  2004-08-28        047XX S KEDZIE AVE  890        THEFT   
3406613     HK456306  2004-06-26  009XX N CENTRAL PARK AVE  820        THEFT   
8002131     HT233595  2011-04-04        043XX S WABASH AVE  820        THEFT   
7903289     HT133522  2010-12-30      083XX S KINGSTON AVE  840        THEFT   
10402076    HZ138551  2016-02-02           033XX W 66TH ST  820        THEFT   

                            DESCRIPTION          LOCATION_DESCRIPTION ARREST  \
ID                                                                             
3512276                   FROM BUILDING            SMALL RETAIL STORE  False   
3406613                  $500 AND UNDER                         OTHER  False   
8002131                  $500 AND UNDER  NURSING HOME/RETIREMENT HOME  False   
7903289   FINANCIAL ID THEFT: OVER $300                     RESIDENCE  False   
10402076                 $500 AND UNDER                         ALLEY  False   

         DOMESTIC  BEAT  DISTRICT  WARD COMMUNITY_AREA_NUMBER FBICODE  \
ID                                                                      
3512276     False   911         9  14.0                  58.0       6   
3406613     False  1112        11  27.0                  23.0       6   
8002131     False   221         2   3.0                  38.0       6   
7903289     False   423         4   7.0                  46.0       6   
10402076    False   831         8  15.0                  66.0       6   

         X_COORDINATE Y_COORDINATE  YEAR   LATITUDE  LONGITUDE  \
ID                                                               
3512276     1155838.0    1873050.0  2004   41.80744 -87.703956   
3406613     1152206.0    1906127.0  2004   41.89828 -87.716406   
8002131     1177436.0    1876313.0  2011  41.815933 -87.624642   
7903289     1194622.0    1850125.0  2010  41.743665 -87.562463   
10402076    1155240.0    1860661.0  2016  41.773455  -87.70648   

                             LOCATION  
ID                                     
3512276     41.8074405, -87.703955849  
3406613   41.898279962, -87.716405505  
8002131   41.815933131, -87.624642127  
7903289   41.743665322, -87.562462756  
10402076  41.773455295, -87.706480471

In [36]:
df_by_elt.head()

CASE_NUMBER       DATE                     BLOCK IUCR PRIMARY_TYPE  \
ID                                                                            
3512276     HK587712 2004-08-28        047XX S KEDZIE AVE  890        THEFT   
3406613     HK456306 2004-06-26  009XX N CENTRAL PARK AVE  820        THEFT   
8002131     HT233595 2011-04-04        043XX S WABASH AVE  820        THEFT   
7903289     HT133522 2010-12-30      083XX S KINGSTON AVE  840        THEFT   
10402076    HZ138551 2016-02-02           033XX W 66TH ST  820        THEFT   

                            DESCRIPTION          LOCATION_DESCRIPTION  ARREST  \
ID                                                                              
3512276                   FROM BUILDING            SMALL RETAIL STORE   False   
3406613                  $500 AND UNDER                         OTHER   False   
8002131                  $500 AND UNDER  NURSING HOME/RETIREMENT HOME   False   
7903289   FINANCIAL ID THEFT: OVER $300                     RESIDENCE   False   
10402076                 $500 AND UNDER                         ALLEY   False   

          DOMESTIC  BEAT  DISTRICT  WARD COMMUNITY_AREA_NUMBER FBICODE  \
ID                                                                       
3512276      False   911         9  14.0                  58.0       6   
3406613      False  1112        11  27.0                  23.0       6   
8002131      False   221         2   3.0                  38.0       6   
7903289      False   423         4   7.0                  46.0       6   
10402076     False   831         8  15.0                  66.0       6   

         X_COORDINATE Y_COORDINATE  YEAR   LATITUDE  LONGITUDE  \
ID                                                               
3512276     1155838.0    1873050.0  2004   41.80744 -87.703956   
3406613     1152206.0    1906127.0  2004   41.89828 -87.716406   
8002131     1177436.0    1876313.0  2011  41.815933 -87.624642   
7903289     1194622.0    1850125.0  2010  41.743665 -87.562463   
10402076    1155240.0    1860661.0  2016  41.773455  -87.70648   

                             LOCATION  
ID                                     
3512276     41.8074405, -87.703955849  
3406613   41.898279962, -87.716405505  
8002131   41.815933131, -87.624642127  
7903289   41.743665322, -87.562462756  
10402076  41.773455295, -87.706480471

Run code below to check it

In [46]:
# Dont modify
df_your_etl = pd.read_sql(your_etl_query, con=con)
try:
  if df_your_etl.equals(df_by_elt):
    print("Congrats its same")
  else:
    print("Its not same, try again")
except Exception as e:
  print("Your SQL code is error, please check it again", e)

Its not same, try again


In [44]:
df_your_etl.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 533 entries, 3512276 to 1326195
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   CASE_NUMBER            533 non-null    object
 1   DATE                   533 non-null    object
 2   BLOCK                  533 non-null    object
 3   IUCR                   533 non-null    object
 4   PRIMARY_TYPE           533 non-null    object
 5   DESCRIPTION            533 non-null    object
 6   LOCATION_DESCRIPTION   533 non-null    object
 7   ARREST                 533 non-null    object
 8   DOMESTIC               533 non-null    object
 9   BEAT                   533 non-null    int64 
 10  DISTRICT               533 non-null    int64 
 11  WARD                   533 non-null    object
 12  COMMUNITY_AREA_NUMBER  533 non-null    object
 13  FBICODE                533 non-null    object
 14  X_COORDINATE           533 non-null    object
 15  Y_COORDINATE 

In [45]:
df_by_elt.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 533 entries, 3512276 to 1326195
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   CASE_NUMBER            533 non-null    object        
 1   DATE                   533 non-null    datetime64[ns]
 2   BLOCK                  533 non-null    object        
 3   IUCR                   533 non-null    object        
 4   PRIMARY_TYPE           533 non-null    object        
 5   DESCRIPTION            533 non-null    object        
 6   LOCATION_DESCRIPTION   533 non-null    object        
 7   ARREST                 533 non-null    bool          
 8   DOMESTIC               533 non-null    bool          
 9   BEAT                   533 non-null    int64         
 10  DISTRICT               533 non-null    int64         
 11  WARD                   533 non-null    object        
 12  COMMUNITY_AREA_NUMBER  533 non-null    object        
